### PR TITLE
.downcase(Unicode::CaseOptions::Fold) using char ranges

### PR DESCRIPTION
**Pull request by akzhan**
The Fold option is required for correct Unicode string comparison using fold case.

Reimplementation of #4511 using char ranges.

Less memory used.

### CODE REVIEW

File: src/unicode/unicode.cr [hunk 1]
```diff
@@ -141,6 +153,16 @@ module Unicode
     nil
   end
 
+  private def self.check_downcase_fold(char, options)
+    if options.fold?
+      result = search_ranges(casefold_ranges, char.ord)
+      return [char.ord + result] if result
```
Comment (asterite) [inline src/unicode/unicode.cr]: Use a tuple to avoid a memory allocation:

```crystal
return {char.ord + result} if result
```

**Comment by akzhan** [comment]
String.compare will be reworked by another PR

**Comment by akzhan** [comment]
Is it GTG? 

I need these changes to continue Unicode support.

**Comment by asterite** [comment]
@akzhan I'm not sure why this is useful. Where would one use it? Plus, if it's to solve another issue (the one about unicode normalization) I would send a single PR targeting that. Is there any other language where you can downcase and specify to fold case?

**Comment by akzhan** [comment]
@asterite full Unicode case mapping added in Ruby 2.4 with `.downcase(:fold)`.

Real case fold comparison doing like:

```python
def NFD(text):
    return unicodedata.normalize('NFD', text)

def canonical_caseless(text):
    return NFD(NFD(text).casefold())
```

**Comment by akzhan** [comment]
`String.normalize` will be subject of another PR.

**Comment by asterite** [comment]
@akzhan Thank you! I'll review this today or during this week and probably merge it.

**Pull request merged**